experimental/sev_guest/src/instructions.rs

```diff
@@ -0,0 +1,193 @@
+//
+
+//! Rust instruction wrappers for managing page state and interacting with the hypervisor.
+
+use bitflags::bitflags;
+use core::arch::asm;
+use strum::FromRepr;
+
+/// Whether a page is in the validated state or not.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum Validation {
+    /// The page is not validated.
+    Unvalidated = 0,
+    /// The page is validated.
+    Validated = 1,
+}
+
+/// The size of a memory page.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum PageSize {
+    /// The page is a 4KiB page.
+    Page4KiB = 0,
+    /// The page is a 2MiB page.
+    Page2MiB = 1,
+}
+
+/// The potential errors when calling the PVALIDATE or RMPADJUST instructions.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum InstructionError {
+    /// The input parameters were invalid.
+    FailInput = 1,
+    /// Insufficient permissions.
+    FailPermission = 2,
+    /// The page size does not match the page size entry in the RMP.
+    FailSizeMismatch = 6,
+    /// The page validation status was not updated. This value is software defined and will not be
+    /// returned by the hardware instruction.
+    ValdationStatusNotUpdated = 255,
+}
+
+/// Marks a page as validated or unvalidated in the RMP.
+///
+/// See the PVALIDATE instruction in <https://www.amd.com/system/files/TechDocs/24594.pdf> for more details.
+#[inline]
+pub fn pvalidate(
+    page_addr: u64,
+    page_size: PageSize,
+    validated: Validation,
+) -> Result<(), InstructionError> {
+    let page_size = page_size as u32;
+    let validated = validated as u32;
+    let result: u32;
+    let carry: u8;
+    // Safety: this call does not modify the guest memory contents, so does not violate memory
+    // safety.
+    unsafe {
+        asm!(
+            "pvalidate",
+            "setc dl",
+            in("rax") page_addr,
+            in("ecx") page_size,
+            in("edx") validated,
+            lateout("eax") result,
+            lateout("dl") carry,
+            options(nomem, nostack)
+        );
+    }
+    if result == 0 {
+        if carry == 0 {
+            Ok(())
+        } else {
+            // If the carry flag is not 0, it indicates that the validated state was not changed.
+            Err(InstructionError::ValdationStatusNotUpdated)
+        }
+    } else {
+        Err(InstructionError::from_repr(result)
+            .expect("Invalid return value from PVALIDATE instruction."))
+    }
+}
+
+bitflags! {
+    /// Permission mask used by the RMP.
+    ///
+    /// See the RMPADJUST instruction in <https://www.amd.com/system/files/TechDocs/24594.pdf> for more details.
+    pub struct PermissionMask: u8 {
+        /// The target VMPL can read the page.
+        const READ = 1;
+        /// The target VMPL can write to the page.
+        const WRITE = 2;
+        /// Code in the page can be executed in ring 3.
+        const EXECUTE_USER = 4;
+        /// Code in the page can be executed in rings 0..2.
+        const EXECUTE_SUPERVISOR = 8;
```

Comment:
Where do these come from? :)

I looked at the linked do's `RMPADJUST`, and I assume they refer to the `TARGET_PERM_MASK` (Desired permission mask settings) field? I couldn't find a definition for the values of that field though

Reply:
Good question. The details are actually in a different doc (volume 2 of the manual). Updated the doc comment to reference the other doc.